cli/data.go

```diff
@@ -35,89 +37,182 @@ func (c *AppClient) BinaryData(dst string, filter *datapb.Filter) error {
+		resp, err := client.BinaryDataByFilter(ctx, &datapb.BinaryDataByFilterRequest{
+			DataRequest: &datapb.DataRequest{
+				Filter: filter,
+				Limit:  uint64(limit),
+				Last:   last,
+			},
+			CountOnly: false,
+		})
 		if err != nil {
 			return err
 		}
-		if _, err := jsonFile.Write(mdJSONBytes); err != nil {
-			return err
+		// If no data is returned, there is no more data.
+		if len(resp.GetData()) == 0 {
+			return nil
 		}
+		last = resp.GetLast()
 
-		gzippedBytes := datum.GetBinary()
-		r, err := gzip.NewReader(bytes.NewBuffer(gzippedBytes))
-		if err != nil {
-			return err
+		for _, bd := range resp.GetData() {
+			ids <- bd.GetMetadata().GetId()
 		}
+	}
+}
 
-		//nolint:gosec
-		dataFile, err := os.Create(filepath.Join(dst, dataDir, fileName+datum.GetMetadata().GetFileExt()))
-		if err != nil {
-			return errors.Wrapf(err, fmt.Sprintf("error creating file for file %s", datum.GetMetadata().GetId()))
-		}
-		//nolint:gosec
-		if _, err := io.Copy(dataFile, r); err != nil {
-			return err
-		}
-		if err := r.Close(); err != nil {
-			return err
-		}
-		numFilesDownloaded++
-		if numFilesDownloaded%logEveryN == 0 {
-			fmt.Fprintf(c.c.App.Writer, "downloaded %d files\n", numFilesDownloaded)
+func downloadBinary(ctx context.Context, client datapb.DataServiceClient, dst, id string) error {
+	var resp *datapb.BinaryDataByIDsResponse
+	var err error
+	for count := 0; count < maxRetryCount; count++ {
+		resp, err = client.BinaryDataByIDs(ctx, &datapb.BinaryDataByIDsRequest{
+			FileIds:       []string{id},
+			IncludeBinary: true,
+		})
+		if err == nil {
+			break
 		}
 	}
-	if numFilesDownloaded%logEveryN != 0 {
-		fmt.Fprintf(c.c.App.Writer, "downloaded %d files to %s\n", numFilesDownloaded, dst)
+	if err != nil {
+		return errors.Wrapf(err, "received error from server")
 	}
-	return nil
-}
+	data := resp.GetData()
 
-func (c *AppClient) sendBinaryDataByFilterRequest(filter *datapb.Filter, last string) (*datapb.BinaryDataByFilterResponse, error) {
-	return c.dataClient.BinaryDataByFilter(context.Background(), &datapb.BinaryDataByFilterRequest{
-		DataRequest: &datapb.DataRequest{
-			Filter: filter,
-			Limit:  1,
-			Last:   last,
-		},
-		IncludeBinary: true,
-		CountOnly:     false,
-	})
+	if len(data) != 1 {
+		return errors.Errorf("expected a single response, received %d", len(data))
+	}
+
+	datum := data[0]
+	mdJSONBytes, err := protojson.Marshal(datum.GetMetadata())
+	if err != nil {
+		return err
+	}
+
+	//nolint:gosec
+	jsonFile, err := os.Create(filepath.Join(dst, "metadata", datum.GetMetadata().GetId()+".json"))
+	if err != nil {
+		return err
+	}
+	if _, err := jsonFile.Write(mdJSONBytes); err != nil {
+		return err
+	}
+
+	gzippedBytes := datum.GetBinary()
+	r, err := gzip.NewReader(bytes.NewBuffer(gzippedBytes))
+	if err != nil {
+		return err
+	}
+
+	//nolint:gosec
+	dataFile, err := os.Create(filepath.Join(dst, "data", datum.GetMetadata().GetId()+datum.GetMetadata().GetFileExt()))
```

Comment:
Looks like this section also lost the newest changes with dataDir/metadataDir and prepending timestamp to the filename. The latter is especially needed now that we're interleaving downloads :)

If you could locally test a scenario where you're moving a motor while collecting data and, post-sort, seeing all the data in chronological order, that'd be great.

In the future when we support downloads via our cloud service, we can actually sort + package nicely before downloading to browser.

Reply:
Oops yup bad merging, added the naming stuff back.

> If you could locally test a scenario where you're moving a motor while collecting data and, post-sort, seeing all the data in chronological order, that'd be great.

It looks like we weren't sorting tabular data by timestamp before, and this doesn't do tabular export in parallel. Did you mean for binary?

While exporting tabular I do think I found a bug in how we're persisting time stamps for tabular data, though. For any `syncInterval`, it looks like we record the same time requested/received for all data points. You can see this by running this command:
`go run go.viam.com/rdk/cli/cmd data --location_ids=zqegrpxfhl --robot_name=Detector --org_ids=1c614556-2ff9-4234-9a94-d59b0a6d3378 --data_type=tabular --start=2022-11-01T04:00:00.000Z`

This data was being collected at 10hz with a sync interval of 1 minute. If you open the file, you can see every 600 messages have the same time requested/received. I think we should prioritize fixing this.

Reply:
Ah yup found the bug, will tag you on follow up PR

Reply:
Fix out here https://github.com/viamrobotics/app/pull/962/files

Reply:
> It looks like we weren't sorting tabular data by timestamp before, and this doesn't do tabular export in parallel. Did you mean for binary?

Yep meant to add this in the binary section, and both tabular/binary needed merge updates.

> This data was being collected at 10hz with a sync interval of 1 minute. If you open the file, you can see every 600 messages have the same time requested/received. I think we should prioritize fixing this.

Really cool seeing the E2E manual testing uncovering a bug like this. Fantastic job on both for spotting it and pushing out a super quick fix!

Interesting to see that even though it had full test coverage, the tests were expecting the wrong thing. Worth listing out scenarios in PR descriptions to help get high-level consensus on test behavior, since sometimes harder to spot when writing/reviewing as code.